# ICU Deep Learning

## 1. Prior to Using this Notebook

- Review the [ICU Deep Learning project README](https://github.com/duanegoodner/icu-deep-learning/tree/main)
- Follow instructions in [SETUP.md](https://github.com/duanegoodner/icu-deep-learning/blob/main/SETUP.md) for setting up and and running the necessary Docker containers.

## 2. Project Structure

Although this project does not include standard package metadata files (e.g., `pyproject.toml`), the code under `./src/` is organized as a top-level package, **`lstm_adversarial_attack`**, with the following sub-packages:  
- **`query_db`** – Manages interactions with the MIMIC-III PostgreSQL database.  
- **`tuning_db`** – Handles PostgreSQL databases used for **Optuna** hyperparameter tuning.  
- **`preprocess`** – Converts **SQL query output** into numerical arrays used as model inputs and targets.  
- **`dataset`** – Wraps preprocessed arrays into **PyTorch Dataset** objects for use in LSTM models.  
- **`model`** – Conducts **hyperparameter tuning, training, and evaluation** of predictive models.  
- **`attack`** – Tunes and trains an **adversarial attack model** on a subset of data, then uses it to attack the full dataset.  
- **`attack_analysis`** – Organizes and analyzes adversarial attack results, characterizing distributions of discovered adversarial examples.  

## 3. Quick Start: Running from the Command Line

This section provides **minimal** instructions for running the full project pipeline **from the command line** with default settings. It is intended for users who want to execute all steps **as quickly as possible** inside the `lstm_aa_app` container.

For **detailed explanations** of each step, including intermediate analysis and discussion, refer to **subsequent sections** of this notebook, where each step is explored in depth.



## 3.1 Pre-requisites

Before proceeding, ensure that:

✅ You have completed Steps 1–6 in SETUP.md.

✅ You are running a shell inside the lstm_aa_app container.

If you're unsure, refer to [SETUP.md](https://github.com/duanegoodner/icu-deep-learning/blob/main/SETUP.md) for instructions on running and entering.

## 3.2 Running the Full Pipeline with Default Settings  

From a command prompt running inside the container, navigate to:

 ```
 /home/devspace/project/src/lstm_adversarial_attack
 ```
 Then, run the commands shown in the table below. These commands execute all steps in the project pipeline using **default parameters**.  The output paths are relative to `/home/devspace/project/data/` which is mapped by **docker compose** to `./data/` in the local `icu-deep-learning` repository. 

| Step | Command                             | Effects                                                                                     | Output Dir  |
|------|-------------------------------------|---------------------------------------------------------------------------------------------|--------------------------|
| 1    | `python -m query_db`                | Runs all MIMIC-III PostgreSQL queries and saves output as **.csv** files.                   | `query_db/`             |
| 2    | `python -m preprocess`              | Converts database query output into arrays for LSTM model inputs<br>and prediction targets. Saves at **5 intermediate steps**. | `preprocess/`           |
| 3    | `python model/tune_new.py`          | Performs **hyperparameter tuning** for the LSTM predictive model.                           | `model/tuning/`         |
| 4    | `python model/train.py`        | Trains and tests the LSTM model using **selected hyperparameters**.                         | `model/cross_validation/` |
| 5    | `python attack/tune_attack_new.py`  | Tunes attack model hyperparameters for adversarial attacks on a **trained predictive model**. | `attack/tuning/`       |
| 6    | `python attack/attack.py`           | Runs **adversarial attacks** using the tuned attacker.                                      | `attack/frozen_hyperparameter_attack/` |
| 7    | `python -m attack_analysis`         | Analyzes and plots results of adversarial attacks.                                          | `attack/attack_analysis/` |


## 3.3 Reviewing Results  

After running the pipeline, plots characterizing the discovered **adversarial examples** can be found in:  

📂 `./data/attack/attack_analysis/`  

These plots provide detailed characteristics of **attack results** but do **not** include information on the preceding pipeline stages (e.g., model training, preprocessing). For further details on those steps, refer to the respective sections later in this notebook.


## 4. Running Code from within this Notebook

### 4.1 Pre-requisites

The remainder of this notebook contains code cells for running the project in a connected IPython interpreter. Before proceeding, ensure that:

✅ You have completed Steps 1–7 in [SETUP.md](https://github.com/duanegoodner/icu-deep-learning/blob/main/SETUP.md).

✅ This notebook is connected to a Jupyter server instance running inside Docker container `lstm_aa_app`.

If you're unsure, refer to [SETUP.md](https://github.com/duanegoodner/icu-deep-learning/blob/main/SETUP.md) for instructions on how to run Jupyter inside the container.



### 4.2 Check the Python Interpreter and IPython Kernel

In [1]:
!which python
# Output should be: /home/devspace/env/bin/python

/home/devspace/env/bin/python


In [2]:
from IPython.core.getipython import get_ipython
get_ipython().kernel.config["IPKernelApp"]["connection_file"]
# Outptut should be similar to: '/home/gen_user/.local/share/jupyter/runtime/kernel-v2-26202uI0Vk7x2nHkK.json'

'/home/gen_user/.local/share/jupyter/runtime/kernel-v3e244e5730daa0cada4896ee8ab05ebfc2f09f5c0.json'

### 2.3 Test Database Connections

Run the following cell to test the MIMIC-III database:

In [3]:
!python /home/devspace/project/src/lstm_adversarial_attack/query_db/test_mimiciii_db.py

Successfully connected to MIMIC-III database.
Connection to MIMIC-III database successfully closed.


Then run test queries on the databases that will handle hyperparameter tuning data:

In [4]:
!python /home/devspace/project/src/lstm_adversarial_attack/tuning_db/test_tuning_study_dbs.py

model_tuning database successfully queried.
Found 2 tuning studies.
attack_tuning database successfully queried.
Found 3 tuning studies.


### 2.4 Check for GPU

The PyTorch code in our project will run much faster on a GPU than it will on a CPU. Let's find out if we have GPU access:

In [5]:
import torch
torch.cuda.is_available()

True

### 2.5 Change Working Directory
Many of the code cells in this notebook use relative paths and assume we are in directory `/home/devspace/project/src/lstm_adversarial_attack`, so let's change to that directory.

In [6]:
import os
os.chdir("/home/devspace/project/src/lstm_adversarial_attack")
!pwd

/home/devspace/project/src/lstm_adversarial_attack


## 3. Project Structure
Our `docker-compose.yml` maps the local project root directory to `/home/devspace/project` in the container. Run the following cell for an overview of our project layout. 

In [7]:
!tree -L 1 /home/devspace/project

/home/devspace/project
├── README.md
├── SETUP.md
├── config.toml
├── data
├── docker
├── docs
├── logs
├── notebooks
└── src

6 directories, 3 files


### 3.1 `src/`
The contents of `/home/devspace/project/src/lstm_adversarial_attack` are:

In [8]:
!tree -d -L 1 /home/devspace/project/src/lstm_adversarial_attack

/home/devspace/project/src/lstm_adversarial_attack
├── __pycache__
├── attack
├── attack_analysis
├── config
├── dataset
├── model
├── preprocess
├── query_db
├── tuning_db
└── utils

10 directories


 Code in the sub-directories listed above forms our project pipeline: 
 * **query_db** runs .sql queries to extract patient lab, vital sign, and in-hospital mortality data from the MIMIC-III PostgreSQL database.
 * **preprocess** transforms .sql query output into a form that can be input to PyTorch models. 
 * **model** tunes and trains a PyTorch model for predicting in-hospital mortality based on lab and vital sign time-series data.
 * **attack** tunes and trains a PyTorch attack model that generates adversarial examples for the predictive model.
 * **attack_analysis** generates plots for visualizing characteristics of adversarial examples found by the attack model.

### 3.2 `config.toml`
Project configuration variables are set in the `config.toml` file. We can use the `get_config_value` and `set_config_value` from `utils.notebook_helpers` to read and write to the `config.toml` file.

In [9]:
import utils.notebook_helpers as nh

help(nh.get_config_value)
help(nh.set_config_value)

Help on function get_config_value in module utils.notebook_helpers:

get_config_value(config_key: str) -> str
    Gets value from config.toml
    :param config_key: config.toml key as dotted string
    :return: value corresponding to config.toml key

Help on function set_config_value in module utils.notebook_helpers:

set_config_value(config_key: str, value: Any)
    Sets value from config.toml
    :param config_key: config.toml key as dotted string
    :param value: value to assign to key
    :return: None



Here is a quick example of reading and writing to `config.toml`:

In [10]:
orig_kfold_random_seed = nh.get_config_value("model.tuner_driver.kfold_random_seed")
print(f"Original value: {orig_kfold_random_seed}")

nh.set_config_value("model.tuner_driver.kfold_random_seed", 2024)
modified_kfold_random_seed = nh.get_config_value("model.tuner_driver.kfold_random_seed")
print(f"Value changed to: {modified_kfold_random_seed}")

nh.set_config_value("model.tuner_driver.kfold_random_seed", orig_kfold_random_seed)

final_kfold_random_seed = nh.get_config_value("model.tuner_driver.kfold_random_seed")
print(f"Final value: {final_kfold_random_seed}")

Original value: 1234
Value changed to: 2024
Final value: 1234


### 3.3 `data/`

For each of the critical directories under `src/lstm_adversarial_attack/`, there is a corresponding directory under `data/`

In [11]:
!tree -L 1 /home/devspace/project/data

/home/devspace/project/data
├── attack
├── attack_analysis
├── model
├── preprocess
└── query_db

5 directories, 0 files


Subdirectories under `/data` contain output from runs of a particular data-generating sub-package or module. Example contents of `data/query_db` are shown below. There are data from three runs, with each containing `.csv` query result files as well as `.toml` files capturing the project configuration settings at the time of the run.

In [12]:
!tree /home/devspace/project/data/query_db

/home/devspace/project/data/query_db
├── 20250222210125577048
│   ├── configs
│   │   ├── session_config.toml
│   │   └── session_config_paths.toml
│   ├── icustay_detail.csv
│   ├── pivoted_bg.csv
│   ├── pivoted_lab.csv
│   └── pivoted_vital.csv
└── 20250223131513840062
    ├── configs
    │   ├── session_config.toml
    │   └── session_config_paths.toml
    ├── icustay_detail.csv
    ├── pivoted_bg.csv
    ├── pivoted_lab.csv
    └── pivoted_vital.csv

4 directories, 12 files


### 3.4 Keeping Track of the Data Pipeline when Running in a Notebook
We will use an instance of `notebook_helpers.PipelineInfo` to store and retrieve module and sub-package runs' metadata.

In [13]:
help(nh.PipelineInfo)

Help on class PipelineInfo in module utils.notebook_helpers:

class PipelineInfo(builtins.object)
 |  PipelineInfo(sessions: dict[str, utils.notebook_helpers.SessionInfo] = None, next_session_index: int = 1)
 |  
 |  Container metadata of data-generating sessions. Intended for use in Jupyter
 |   notebooks.
 |  
 |  Methods defined here:
 |  
 |  __init__(self, sessions: dict[str, utils.notebook_helpers.SessionInfo] = None, next_session_index: int = 1)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  get_stored_session(self, session_type: utils.notebook_helpers.SessionType, session_id: str | int = None) -> utils.notebook_helpers.SessionInfo
 |      Gets info on a stored session. If PipelineInfo object has more than one
 |      entry for session of session_type, must specify session ID.
 |      :param session_type: Type of session to retrieve
 |      :param session_id: ID of session
 |      :return: info for session
 |  
 |  store_session(self, session_typ

For now, we will instantiate an empty PipelineInfo object.

In [15]:
import pprint
pipeline_info = nh.PipelineInfo()
pprint.pprint(pipeline_info.sessions)

{}


Each time we run any data-generating code, we will store metadata in our PipelineInfo object. Note that we can also use the `store_session` method to store metadata to a data-generating run performed outside of our notebook session. This outside run could have been initiated from the command line, or from an ealier notebook session.

## 4. Database Queries

Raw ICU patient data can be extracted from the MIMIC-III database using modified versions of four `.sql`queries from the [MIT-LCP mimic-code repository](https://github.com/MIT-LCP/mimic-code/tree/main/mimic-iii/concepts/pivot).

### 4.1 Running the Queries

We connect to the database and execute the queries by running the [\_\_main__](../src/lstm_adversarial_attack/query_db/__main__.py) module of the [query_db](../src/lstm_adversarial_attack/query_db/\_\_init__.py) sub-package. 

In [16]:
!python -m query_db --help

usage: __main__.py [-h] [-q [QUERY_DIR]]

Runs .sql queries on MIMIC-III database

options:
  -h, --help            show this help message and exit
  -q [QUERY_DIR], --query_dir [QUERY_DIR]
                        Directory containing .sql query files. Defaults to
                        path specified by paths.db.output_root in config.toml


In [16]:
!python -m query_db

Starting query session: 20250223131513840062

Query 1 of 4
Executing: /home/devspace/project/src/lstm_adversarial_attack/query_db/mimiciii_queries/icustay_detail.sql
Done. Query time = 0.25 seconds
Writing result to csv: /home/devspace/project/data/query_db/20250223131513840062/icustay_detail.csv
Done. csv write time = 0.18 seconds

Query 2 of 4
Executing: /home/devspace/project/src/lstm_adversarial_attack/query_db/mimiciii_queries/pivoted_bg.sql
Done. Query time = 8.50 seconds
Writing result to csv: /home/devspace/project/data/query_db/20250223131513840062/pivoted_bg.csv
Done. csv write time = 1.46 seconds

Query 3 of 4
Executing: /home/devspace/project/src/lstm_adversarial_attack/query_db/mimiciii_queries/pivoted_lab.sql
Done. Query time = 10.67 seconds
Writing result to csv: /home/devspace/project/data/query_db/20250223131513840062/pivoted_lab.csv
Done. csv write time = 2.41 seconds

Query 4 of 4
Executing: /home/devspace/project/src/lstm_adversarial_attack/query_db/mimiciii_queries

### 4.2 Storing Query Session Info
We can register the most recently created query session in our `PipelineInfo` storage container using:

In [17]:
pipeline_info.store_session(session_type = nh.SessionType.DB_QUERIES)

20250223131513840062 stored


In [18]:
import pprint
pprint.pprint(pipeline_info.sessions)

{'20250223131513840062': SessionInfo(session_type=<SessionType.DB_QUERIES: 1>,
                                     session_id='20250223131513840062',
                                     comment=None)}


## 5. Preprocess

### 5.1 Implementation Details

We will use the [`preprocess`](../src/lstm_adversarial_attack/preprocess/__init__.py) sub-package to transform information from the `.csv` files output by the `.sql` queries into numpy arrays (which can then be easily converted into PyTorch tensors). Running this sub-package instantiates a `Preprocessor` object with a `.preprocess_modules` attribute assigned by the following code in  [`preprocessor.py`](../src/lstm_adversarial_attack/preprocess/preprocessor.py):

```
self.preprocess_modules = [
            prf.Prefilter(),
            imc.ICUStayMeasurementCombiner(),
            slb.FullAdmissionListBuilder(),
            fb.FeatureBuilder(),
            ff.FeatureFinalizer(),
        ]
```
Each element of the `.preprocess_modules` attribute is a subclass of [`PreprocessModule`](../src/lstm_adversarial_attack/preprocess/preprocess_module.py).

* [`Prefilter`](../src/lstm_adversarial_attack/preprocess/prefilter.py) reads the database query outputs into Pandas Dataframes, removes all data related to patients younger than 18 years in age, ensures consistent column naming formats, and takes care of datatype details.
* [`ICUStayMeasurementCombiner`](../src/lstm_adversarial_attack/preprocess/icustay_measurement_combiner.py) performs various joins (aka "merges" in the language of Pandas) to combine lab and vital sign measurement data with ICU stay data.
* [`FullAdmissionListBuilder`](../src/lstm_adversarial_attack/preprocess/sample_list_builder.py) generates a list consisting of one FullAdmissionData object per ICU stay. The attributes of a FullAdmissionData object include ICU stay info, and a dataframe containing the measurement and timestamp data for all vital sign and lab data associated with the ICU stay.
* [`FeatureBuilder`](../src/lstm_adversarial_attack/preprocess/feature_builder.py) resamples the time series datafame to one-hour intervals, imputes missing data, winsorizes measurement values (with cutoffs at the 5th and 95th global percentiles), and normalizes the measuremnt values so all data are between 0 and 1.
* [`FeatureFinalizer`](../src/lstm_adversarial_attack/preprocess/feature_finalizer.py) selects the data observation time window (default starts at hospital admission time and ends 48 hours after admission). This module outputs the entire dataset features as a list of numpy arrays, and the mortality labels as a list of integers. These data structures (saved as .pickle files) will be convenient starting points when the `tune_train` and `attack` sub-packages need to create PyTorch Datasets.

### 5.2 Set Preprocess Config Values

In [19]:
nh.set_config_value("preprocess", {'min_age': 18,
 'min_los_hospital': 1,
 'min_los_icu': 1,
 'bg_data_cols': ['potassium', 'calcium', 'ph', 'pco2', 'lactate'],
 'lab_data_cols': ['albumin',
  'bun',
  'creatinine',
  'sodium',
  'bicarbonate',
  'platelet',
  'glucose',
  'magnesium'],
 'vital_data_cols': ['heartrate',
  'sysbp',
  'diasbp',
  'tempc',
  'resprate',
  'spo2'],
 'winsorize_low': '5%',
 'winsorize_high': '95%',
 'resample_interpolation_method': 'linear',
 'resample_limit_direction': 'both',
 'min_observation_hours': 48,
 'observation_window_hours': 48,
 'observation_window_start': 'intime'})

### 5.3 Run the Preprocess Modules

We can run all preprocessing modules by executing the `preprocess` sub-packages `__main__`

In [20]:
!python -m preprocess --help

usage: __main__.py [-h] [-d [DB_RESULT_ID]]

Takes data output from database query, and runs through preprocess modules.

options:
  -h, --help            show this help message and exit
  -d [DB_RESULT_ID], --db_result_id [DB_RESULT_ID]
                        ID of database query session to take output from.
                        Defaults to latest query.


In [21]:
db_queries_session = pipeline_info.get_stored_session(session_type=nh.SessionType.DB_QUERIES)

Retrieved session: 20250223131513840062 


In [22]:
!python -m preprocess -d {db_queries_session.session_id}

Preprocess session will use data from database query session 20250223131513840062
Starting preprocess session 20250223131653849429

Running Prefilter
Prefilter init time = 5.918374300003052
Prefilter process time = 1.6608531475067139
Prefilter export time = 0.3108096122741699
Output saved in /home/devspace/project/data/preprocess/20250223131653849429/1_prefilter

Running ICUStayMeasurementMerger
ICUStayMeasurementMerger init time = 0.01996326446533203
ICUStayMeasurementMerger process time = 14.454586029052734
ICUStayMeasurementMerger export time = 1.337817668914795
Output saved in /home/devspace/project/data/preprocess/20250223131653849429/2_merged_stay_measurements

Running AdmissionListBuilder
AdmissionListBuilder init time = 0.0076215267181396484
AdmissionListBuilder process time = 19.395607233047485
AdmissionListBuilder export time = 17.45841383934021
Output saved in /home/devspace/project/data/preprocess/20250223131653849429/3_full_admission_list

Running FeatureBuilder
FeatureBui

In [22]:
pipeline_info.store_session(session_type=nh.SessionType.PREPROCESS)

20250223131653849429 stored


In [23]:
pprint.pprint(pipeline_info.sessions)

{'20250223131513840062': SessionInfo(session_type=<SessionType.DB_QUERIES: 1>,
                                     session_id='20250223131513840062',
                                     comment=None),
 '20250223131653849429': SessionInfo(session_type=<SessionType.PREPROCESS: 2>,
                                     session_id='20250223131653849429',
                                     comment=None)}


### 5.4 Summarize Feature Finalizer Output
We can get information about the array shape and value distributions of the preprocessed using the `preprocess` sub-package's `inspect_feature_finalizer` module.

In [24]:
!python preprocess/inspect_feature_finalizer_output.py --help

usage: inspect_feature_finalizer_output.py [-h] [-p [PREPROCESS_ID]]

options:
  -h, --help            show this help message and exit
  -p [PREPROCESS_ID], --preprocess_id [PREPROCESS_ID]
                        ID of preprocess session to use as data source.
                        Defaults to latest session


In [25]:
preprocess_session = pipeline_info.get_stored_session(session_type=nh.SessionType.PREPROCESS)

Retrieved session: 20250223131653849429 


In [26]:
!python preprocess/inspect_feature_finalizer_output.py -p {preprocess_session.session_id}

Summary of output from preprocess session 20250223131653849429

Number of samples = 37832

Time Series Sequence Length Distribution
----------------------------------------
sequence_length     48
count            37832

Measurement Column Counts Distribution
--------------------------------------
num_measurements     19
count             37832

Min value of any element in any feature matrix = 0.0
Max value of any element in any feature matrix = 1.0

Class Labels Distribution
-------------------------
class_label      0     1
count        33991  3841



Each sample in the FeatureFinalizer output is from a unique ICU stay, and consists of a 2D matrix of input features and a binary class label. Each column in a feature matrix corresponds to a particular lab or vital sign measurement, and each row in a feature matrix corresponds to the number of hours elapsed after a patient's hospital admission time. A class label of 1 indicates an in-hospital mortality event.

When preprocessor parameters in `config.toml` are set to default values, the FeatureFinalizer output consists of 37832 samples, and the shape of all input feature arrays is 48 x 19, and approximately 11% of the preprocessed samples have class label = 1. Later, when we tune and train our predictive model, we will use oversampling techniques to deal with the significant class imbalance.

### 5.5 Preprocessing Time

On an Intel i7-13700K CPU, the above preprocessing work takes approximately 3.9 minutes. Achieving the same transformations on the same machine with preprocessing code from [[1](#References)] takes approximately 45 minutes. This time difference is largely due to the fact that the current project preprocess subpackage avoids using  unnecessary loops and relies heavily vectorized Pandas and Numpy operations.

Additional time reduction could be achieved by parellelizing the preprocess computations with tools such as [pandaparallel](https://github.com/nalepae/pandarallel) or [pyspark](https://spark.apache.org/docs/3.3.1/api/python/index.html).

## 6. Model Architecture

The starting point for our predictive model is based on the model in [1] and consists of the following layers:

| Layer # | Description        | Input Shape                            | Parameters          | Output Shape           | Activation       |
| ------- | ------------------ | -------------------------------------- | ------------------- | ---------------------- | ---------------- |
| 1       | Bidirectional LSTM | (b, t<sub>max</sub> = 48, n<sub>meas</sub> = 19) | n<sub>LSTM</sub>    | (b, 2n<sub>LSTM</sub>) | a<sub>LSTM</sub> |
| 2       | Dropoout           | (b, 2n<sub>LSTM</sub>)                 | P<sub>dropout</sub> | (b, 2n<sub>LSTM</sub>) | -                |
| 3       | Fully Connected    | (b, 2n<sub>LSTM</sub>)                 | n<sub>FC</sub>      | (b, n<sub>FC</sub>)    | a<sub>FC</sub>   |
| 4       | Output             | (b, n<sub>FC</sub>)                    | n<sub>out</sub> = 2 | (b, n<sub>out</sub>    | a<sub>out</sub>  |


The parameters from the above table are defined as:

| Parameter           | Description                                             |
| ------------------- | ------------------------------------------------------- |
| b                   | Batch size                                              |
| t<sub>max</sub>     | Maximum input sequence length                           |
| n<sub>meas</sub>    | Number of patient measurement types                     |
| n<sub>LSTM</sub>    | Number of features in a LSTM hidden state               |
| a<sub>LSTM</sub>    | Activation function for the LSTM output                 |
| P<sub>dropout</sub> | Dropout probablity                                      |
| n<sub>FC</sub>      | Numbef of nodes in the fully connected layer            |
| a<sub>FC</sub>      | Activation function for the fully connected layer ouput |
| n<sub>out</sub>     | Number of nodes in the output layer                     |
| a<sub>out</sub>     | Activation function for the output layer                |


Note that n<sub>meas</sub>, n<sub>out</sub>, abd s<sub>max</sub> are fixed. We have chosen to always use all 19 patient measurement types, and our classification problem always has two classes. In our current data pipeline, data collected outside of a specified time window are removed during the final preprocessing phase. If we want the observation window to be tunable, it would be helpful to move the `preprocess.feature_finalizer` module into the `tune_attack` sub-package.

## 7. Model Hyperparameter Tuning

### 7.1 Architectural hyperparameters

The following table lists the ranges architectural parameters to be explored during hyperparameter tuning.

| Parameter           | Tuning Type  | Values                            |
| ------------------- | ------------ | --------------------------------- |
| b                   | Discrete     | 2<sup>k</sup> , k = 5, 6, 7, 8    |                    
| h<sub>LSTM</sub>    | Discrete     | 2<sup>k</sup> , k = 5, 6, 7       |
| a<sub>LSTM</sub>    | Discrete     | ReLU, Tanh                        |
| P<sub>dropout</sub> | Continuous   | 0.000 &mdash; 0.5000        |
| h<sub>FC</sub>      | Discrete     | 2<sup>k</sup> , k = 4, 5, 6, 7, 8 |
| a<sub>FC</sub>      | Discrete     | ReLU, Tanh                        |


### 7.2 Trainer hyperparameters




During hyperparameter tuning, we also explore different training optimization algorithms and learning rates.

| Parameter     | Tuning Type | Values             |
| ------------- | ----------- | ------------------ |
| Optimizer     | Discrete    | SGD, RMSprop, Adam |
| Learning Rate | Continuous  | 1e-5 - 1e-1        |

When using the Adam optimizer, we always use the Pytorch default values of $\beta_1 = 0.9, \beta_2 = 0.999, \epsilon = 10^{-8}$. 

### 7.3 Implementation Details
The [`HyperParameterTuner`](../src/lstm_adversarial_attack/tune_train/hyperparameter_tuner.py) class in the [`model`](../src/lstm_adversarial_attack/model/__init__.py) sub-package implements a cross-validation tuning scheme that utilizes the [Optuna](https://optuna.org/) framework. The boundaries of hyperparameter space to explore during tuning are set in the `[model.tuner_driver.tuning_ranges]` section of the projectr `config.toml` file.

Other model hyperparameter tuning settings are also configured under `[model.tuner_driver]`. In the standard configuration, a PyTorch [`StratifiedKFold`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html) generator is used to assign samples to each fold. When selecting samples for each training batch, we use a [`DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) with a [`WeightedRandomSampler`](https://pytorch.org/docs/stable/data.html#torch.utils.data.WeightedRandomSampler) to oversample from the minority class (label = 1). For a given set of hyperparameters, the [`HyperParameterTuner.objective_fn`](../src/lstm_adversarial_attack/tune_train/hyperparaemter_tuner.py) method returns the mean validation loss across the K folds, and this mean loss is used as a minimization target by an Optuna [`TPESampler`](https://optuna.readthedocs.io/en/stable/reference/samplers/generated/optuna.samplers.TPESampler.html) to select new sets of hyperparameters for additional trials. [`HyperParameterTuner`](../src/lstm_adversarial_attack/tune_train/hyperparaemter_tuner.py) also uses an Optuna [`MedianPruner`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.pruners.MedianPruner.html) to stop unpromising trials early.

### 7.4 Model Tuning Configuration Settings
Set `[model.tuner_driver]` configuration values:

In [27]:
nh.set_config_value("model.tuner_driver", {'num_trials': 60,
 'num_folds': 5,
 'num_cv_epochs': 10,
 'epochs_per_fold': 5,
 'kfold_random_seed': 1234,
 'performance_metric': 'validation_loss',
 'optimization_direction_label': 'minimize',
 'tuning_output_dir': 'data/model/tuning',
 'pruner_name': 'MedianPruner',
 'sampler_name': 'TPESampler',
 'db_env_var_name': 'MODEL_TUNING_DB_NAME',
 'fold_class_name': 'StratifiedKFold',
 'collate_fn_name': 'x19m_collate_fn',
 'cv_mean_tensorboard_metrics': ['accuracy',
  'auc',
  'f1',
  'precision',
  'recall',
  'validation_loss'],
 'tuning_ranges': {'log_lstm_hidden_size': [5, 7],
  'lstm_act_options': ['ReLU', 'Tanh'],
  'dropout': [0.0, 0.5],
  'log_fc_hidden_size': [4, 8],
  'fc_act_options': ['ReLU', 'Tanh'],
  'optimizer_options': ['Adam', 'RMSprop', 'SGD'],
  'learning_rate': [1e-05, 0.1],
  'log_batch_size': [5, 8]},
 'pruner_kwargs': {'n_startup_trials': 5, 'n_warmup_steps': 3},
 'sampler_kwargs': {}})

The `[model.tuner_driver]` secton of the `config.toml` includes parameters that determine the number of tuning trials, cross-validation folds, and epochs. With the values set above, we run an Optuna study with 60 trials. Each trial uses 5-fold cross-validation, and we run `num_cv_epochs * epochs_per_fold = 10 * 5 = 50` total epochs on each fold. (NOTE: Consider changing tname of `epochs_per_fold` to something less confusing.)

### 7.5 Start a New Hyperparameter Tuning Study
Before starting, a few things to note:
* Depending your GPU compute power, running the full 30 trials could take 2 - 20 hours.
* Results will be saved to a newly created directory (with a timestamp-based name) under `data/model/tuning/<tuning_session_id>`. 
* If the study is stopped early (via CTRL-C or the Jupyter Stop button), learning from whatever trials have completed up to that point will be saved.
* While the tuning trials are running, read ahead to the notebook section with instructions on how to monitor progress in Tensorboard.

We can start a new hyperparaemter tuning session by running the `tune_new` module in the `model` sub-package.

In [28]:
!python model/tune_new.py --help

usage: tune_new.py [-h] [-p [PREPROCESS_ID]] [-r]

Creates and runs a new model hyperparameter tuning study.

options:
  -h, --help            show this help message and exit
  -p [PREPROCESS_ID], --preprocess_id [PREPROCESS_ID]
                        ID of preprocess session to use as data source.
                        Defaults to most recently created preprocess session.
  -r, --redirect        Redirect terminal output to log file


Since terminal output during tuning can be very long, we will use the  `-r` option to redirect output to a log file and keep our notebook tidy.

In [29]:
preprocess_session_for_tuning_input = pipeline_info.get_stored_session(session_type=nh.SessionType.PREPROCESS)

Retrieved session: 20250223131653849429 


In [31]:
!python model/tune_new.py -p {preprocess_session_for_tuning_input.session_id} -r

Starting new model hyperparameter tuning session 20250223132531419387

To monitor tuning in tensorboard, run the following command in another terminal:
tensorboard --logdir=/home/devspace/project/data/model/tuning/20250223132531419387/tensorboard --host=0.0.0.0
Then go to http://localhost:6006/ in your browser.


stdout and stderr will be redirected to /home/devspace/project/logs/model/tuning/20250223132531419387.log
Output can be viewed in real time by running the following command in another terminal:
tail -f /home/devspace/project/logs/model/tuning/20250223132531419387.log
^C


The above command will complete once we have run the number of trials specified by `"model.tuner_driver.num_trials` in our `config.toml`, or we can also stop cell execution early.

Next, we store the model tuning session ID in our `PipelineInfo` container.

In [30]:
pipeline_info.store_session(session_type=nh.SessionType.MODEL_TUNING)

20250223132531419387 stored


### 7.6 Resume an Existing Hyperparameter Tuning Study
We can run additional trials for an existing study using the `model` sub-package's `tune_resume` module.

In [31]:
!python model/tune_resume.py --help

usage: tune_resume.py [-h] [-t [MODEL_TUNING_ID]] [-r]

Runs additional model hyperparameter tuning trials for an existing tuning
study. Uses previously saved ModelTunerDriver and optuna.Study as starting
points. New trial results are added to the existing Study.

options:
  -h, --help            show this help message and exit
  -t [MODEL_TUNING_ID], --model_tuning_id [MODEL_TUNING_ID]
                        ID of model tuning session to resume. Defaults to ID
                        of most recently created session.
  -r, --redirect        Redirect terminal output to log file


If we want to continue a study, we can un-comment each of the next two cells, and assign a value to `model_tuning_id_for_continuation`

In [32]:
model_tuning_session_to_continue = pipeline_info.get_stored_session(session_type=nh.SessionType.MODEL_TUNING)

Retrieved session: 20250223132531419387 


In [33]:
!python model/tune_resume.py -t {model_tuning_session_to_continue.session_id} -r

Continuing existing model hyperparameter tuning session 20250223132531419387
To monitor tuning in tensorboard, run the following command in another terminal:
tensorboard --logdir=/home/devspace/project/data/model/tuning/20250223132531419387/tensorboard --host=0.0.0.0' in a different terminal
Then go to http://localhost:6006/ in your browser.


stdout and stderr will be redirected to /home/devspace/project/logs/model/tuning/20250223132531419387.log
Output can be viewed in real time by running the following command in another terminal:
tail -f /home/devspace/project/logs/model/tuning/20250223132531419387.log
^C


### 7.7 Monitor Tuning Progress with Tensorboard

While we are tuning hyperparameters, we can monitor results in Tensorboard. Use Jupyter Lab to open a new terminal, and run:

```
tensorboard --logdir=/home/devspace/project/data/model/tuning/<tuning-session-id>/tensorboard --host=0.0.0.0
```

Then, in your browser, go to: `http://localhost:6006/`. You should see something like the screenshot below.  The x-axis for all plots is epoch number. (Unfortunately, there is no good way to add axis labels in Tensorboard.) Note: `<tuning-session-ID>` is included in the output when running the `tune_new` and/or `tune_resume` modules.

Here is an example screen-shot of plots displayed in Tensorboard.

![tensorboard_image](images/tensorboard_model_tuning_50_epochs.png)

### 7.9 Review Hyperparameters and Objective Function Scores from Model Tuning Study Trial(s)

In [32]:
!python model/view_model_hyperparameters.py --help

usage: view_model_hyperparameters.py [-h] [-t [MODEL_TUNING_ID]]
                                     [--model_tuning_trial_number MODEL_TUNING_TRIAL_NUMBER]

Retrieves and displays hyperparameters tested during a model tuning session.

options:
  -h, --help            show this help message and exit
  -t [MODEL_TUNING_ID], --model_tuning_id [MODEL_TUNING_ID]
                        ID of model tuning session. Defaults to most recently
                        created session.
  --model_tuning_trial_number MODEL_TUNING_TRIAL_NUMBER, -n MODEL_TUNING_TRIAL_NUMBER
                        Optional integer specifying trial number (within
                        study).Defaults to best trial from study.


In [34]:
model_tuning_session_to_review = pipeline_info.get_stored_session(session_type=nh.SessionType.MODEL_TUNING)

Retrieved session: 20250223132531419387 


In [35]:
!python model/view_model_hyperparameters.py -t {model_tuning_session_to_review.session_id}

Summary of model tuning session 20250223132531419387, trial number 0:
Objective function value = 0.4108834130882461
Hyperparameters = 
{'dropout': 0.040370800817631836,
 'fc_act_name': 'ReLU',
 'learning_rate': 0.0007766753211242147,
 'log_batch_size': 5,
 'log_fc_hidden_size': 7,
 'log_lstm_hidden_size': 7,
 'lstm_act_name': 'Tanh',
 'optimizer_name': 'RMSprop'}

The best trial from tuning session 20250223132531419387:
Trial number 0 with objective function value = 0.4108834130882461


If we want to view information from a trial other than the best trial from the study, we can specify the trial number to review:

In [36]:
!python model/view_model_hyperparameters.py -t {model_tuning_session_to_review.session_id} -n 1

Summary of model tuning session 20250223132531419387, trial number 1:
Objective function value = 0.5120636310255477
Hyperparameters = 
{'dropout': 0.4641631336568098,
 'fc_act_name': 'ReLU',
 'learning_rate': 5.0980734177706554e-05,
 'log_batch_size': 7,
 'log_fc_hidden_size': 6,
 'log_lstm_hidden_size': 7,
 'lstm_act_name': 'Tanh',
 'optimizer_name': 'Adam'}

The best trial from tuning session 20250223132531419387:
Trial number 0 with objective function value = 0.4108834130882461


## 8. Model Training
For model hyperparameter tuning described in the previous section, we typically run ~50 epochs per fold (in the interest of reducing compute requirements). Based on the validation loss, AUC, and F1 curves from tuning trials, it appears that predictive performance could be improved by training for a larger number of epochs. We now run another round of Stratified K-fold cross-validation with our best set of parameters with a larger number of epochs.

### 8.1 Notes on our Method
* We are using "flat" cross-validation (as was done in previous studies on this dataset). This method computationally less expensive than nested cross-validation. Flat cross-validation has the potential to overestimate of model performance. In many cases the magnitude of overestimation is small. We also mitigate this effect by using a different set of (randomly generated) fold assignments than was used for hyperparameter tuning. 
* By selecting our hyperparameters based on the smaller number of epochs (100), we favor models that are faster to to train. It is possible that using a larger number of epochs in the tuning runs would have yielded a different (and better) set of "best" hyperparameters, but would also be computationally more expensive.

### 8.2 Cross-Validation Training Settings
Settings used during model training are specified in the `[model.cv_driver_settings]` section of the `config.toml` file.

In [37]:
nh.set_config_value("model.cv_driver_settings", {'collate_fn_name': 'x19m_collate_fn',
 'epochs_per_fold': 1000,
 'eval_interval': 10,
 'fold_class_name': 'StratifiedKFold',
 'kfold_random_seed': 20240807,
 'num_folds': 5,
 'single_fold_eval_fraction': 0.2})

### 8.3 Run Cross-Validation Training
We can begin a training session by running the `train` module in the `model` sub-package.

In [37]:
!python model/train.py --help

usage: train.py [-h] [-t [MODEL_TUNING_ID]] [-r]
                [--model_tuning_trial_number MODEL_TUNING_TRIAL_NUMBER]

Runs cross-validation training using the best model hyperparameters from a
particular model tuning session.

options:
  -h, --help            show this help message and exit
  -t [MODEL_TUNING_ID], --model_tuning_id [MODEL_TUNING_ID]
                        ID of model tuning session that hyperparameters are
                        obtained from. Defaults to the most recently created
                        session.
  -r, --redirect        Redirect terminal output to log file
  --model_tuning_trial_number MODEL_TUNING_TRIAL_NUMBER, -n MODEL_TUNING_TRIAL_NUMBER
                        Optional integer specifying trial number (within
                        study).Defaults to best trial from study.


In [38]:
model_tuning_session_for_training_input = pipeline_info.get_stored_session(session_type=nh.SessionType.MODEL_TUNING)

Retrieved session: 20250223132531419387 


In [41]:
!python model/train.py -t {model_tuning_session_for_training_input.session_id} -r

Starting new cross-validation training session 20250225120514809962.

Will use best hyperparameters from model tuning session 20250223132531419387

To monitor training in tensorboard, run the following command in another terminal:
tensorboard --logdir /home/devspace/project/data/model/cross_validation/20250225120514809962/tensorboard --host=0.0.0.0
Then go to http://localhost:6006/ in your browser.

stdout and stderr will be redirected to /home/devspace/project/logs/model/training/20250225120514809962.log
Output can be viewed in real time by running the following command in another terminal:
tail -f /home/devspace/project/logs/model/training/20250225120514809962.log


When training is complete, store the session ID in our `PipelineInfo` object with:

In [39]:
pipeline_info.store_session(session_type=nh.SessionType.CV_TRAINING)

20250225120514809962 stored


### 8.4 Monitor Cross-Validation Progress in Tensorbard

To view training curves in tensorboard, use Jupyter Lab to open a new terminal, and run:
```
tensorboard --logdir /home/devspace/project/data/model/tuning/<cross-validation-training-session-id>/tensorboard --host=0.0.0.0
```

Then, go to http://localhost:6006 in your browser.

This Tensorboard screenshot was taken at the end of a 5-fold, 1000 epoch per fold cross-validation run.
![tensorboard_image](images/tensorboard_model_training_1000_epochs.png)

### 8.5 Model Training Behavior: Continued Improvement at High Epoch Counts

The above AUC and validation loss curves show continued (though diminishing) improvement in predictive performance during the entire 1000 epochs. The fact that we do not observe any sign of overfitting at such a large number of epochs is somewhat unusual. A likely cause of this behavior is the `WeightedRandomSampler` used in our training `DataLoaders`. Samples with our minority class label (`mortality = 1`) only represent ~15% of the total dataset. To deal with this imbalanced dataset, we oversample from the minority class and undersample from the majority class when creating batches of samples for training. In our current implementation, some samples from the majority class go unseen by the `StandardModelTrainer` for a large number of epochs. The number of unseen samples slowly dwindles (and the amount of information available for training slowly increases), even at very high epoch counts.

### 8.6 Summarize Model Training Results

We can run the `model` sub-package's `view_model_training_summary` module to summarize each fold's best-performing checkpoint as well as the means and standard deviations of performance metrics across all folds.

In [40]:
cv_training_id_for_summary = pipeline_info.get_stored_session(session_type=nh.SessionType.CV_TRAINING)

Retrieved session: 20250225120514809962 


In [41]:
!python model/view_model_training_summary.py --help

usage: view_model_training_summary.py [-h] [-t [CV_TRAINING_ID]]

Displays summary of model training session

options:
  -h, --help            show this help message and exit
  -t [CV_TRAINING_ID], --cv_training_id [CV_TRAINING_ID]
                        ID of cross validation training session to summarize.
                        Defaults to most recently created session


In [42]:
!python model/view_model_training_summary.py -t {cv_training_id_for_summary.session_id}

Summary of Cross Validation Training Session 20250225120514809962

Best Performing Checkpoints by Fold
                          0           1            2           3           4
fold               0.000000    1.000000     2.000000    3.000000    4.000000
epoch            980.000000  980.000000  1000.000000  850.000000  970.000000
train_loss         0.339000    0.343377     0.343347    0.345060    0.347877
validation_loss    0.336766    0.339939     0.340059    0.342779    0.344129
auc                0.976525    0.971363     0.972433    0.969043    0.968571
accuracy           0.976474    0.973302     0.973105    0.970297    0.969107
f1                 0.975886    0.972677     0.972728    0.969886    0.968285
precision          0.987187    0.990155     0.988223    0.989677    0.988503
recall             0.964841    0.955805     0.957712    0.950870    0.948876

Performance Metrics Means and Standard Deviations
                     mean       std
train_loss       0.343732  0.003225
vali

### 8.7 Comparing Training Results with Prior Studies' Predictive Models

The table below compares the predictive performance of the LSTM model in this work with other LSTM-based models using the same dataset. The current model shows the best predictive performance among all models in the table based on AUC and F1 scores. 


| # | Authors     | Model                   | Input Features                                 | AUC             | F1              | Precision       | Recall          |
|---|-------------|-------------------------|------------------------------------------------|-----------------|-----------------|-----------------|-----------------|
|1  | Sun et al.  | LSTM-128 + FC-32 + FC-2 | [13 labs, 6 vitals] x 48 hr                    | 0.9094 (0.0053) | 0.5429 (0.0194) | 0.4100 (0.0272) | 0.8071 (0.0269) |
|2  | Tang et al. | LSTM-256 + FC-2         | [13 labs, 6 vitals] x 48 hr + demographic data | 0.949 (0.003)   | 0.623 (0.012)   | -               | -               |
|3  | Tang et al. | CNN + LSTM-256 + FC-2   | [13 labs, 6 vitals] x 48 hr + demographic data | 0.940 (0.0071)  | 0.633 (0.031)   | -               | -               |
|4  | Tang et al. | CNN + LSTM-256 + FC-2   | [13 labs, 6 vitals] x 48 hr                    | 0.933 (0.006)   | 0.587 (0.025)   | -               | -               |
|5  | Tang et al. | LSTM-256 + FC-2         | [13 labs, 6 vitals] x 48 hr                    | 0.907 (0.006)   | 0.526 (0.013)   | -               | -               |
|6  | This work   | LSTM-128 + FC-16 + FC-2 | [13 labs, 6 vitals] x 48 hr                    | 0.9657 (0.0035) | 0.9669 (0.0038) | 0.9888 (0.0009) | 0.9459 (0.0072) |

> **Notes**
> - The values for the current study were taken from an earlier run of project code and are not "live-populated" when running this notebook.
> - LSTM-X indicates an LSTM with X hidden layers.
> - FC-X indicates a fully connected layer with an output size of X.
> - All LSTMs are bidirectional.
> - The demographic data used in studies #2 and #3 was obtained from MIMIC-III.
 

## 9 Attack Hyperparameter Tuning

Before running an attack on the entire dataset, we tune attack hyperparameters with help from `optuna`. Our approach here is not as rigourous as the one we used for predictive model tuning. We use only a fraction of the total dataset for tuning, and do not perform cross-validation.

## 9.1 Set ID of Training ID to Use for Attack Tuning

For now, we will use the same training session that we summarized in Section 8.6, but this can be changed. 

In [45]:
cv_training_id_for_attack_tuning = cv_training_id_for_summary
cv_training_id_for_attack_tuning.session_id

'20250225120514809962'

### 9.1 Attack Hyperparameter Tuning Config Settings

Settings that affect attack hyperparameter tuning are under `[attack.tuning.ranges]` and `[attack.tuner_driver_settings]`in the `config.toml`. We can view the current values of these settings using:

In [46]:
nh.set_config_value("attack.tuning.ranges", {'kappa': [0.0, 2.0],
 'lambda_1': [1e-07, 1.0],
 'learning_rate': [1e-05, 1.0],
 'log_batch_size': [5, 7],
 'optimizer_options': ['Adam', 'RMSprop', 'SGD']})

nh.set_config_value("attack.tuner_driver_settings", {'db_env_var_name': 'ATTACK_TUNING_DB_NAME',
 'num_trials': 75,
 'epochs_per_batch': 1000,
 'max_num_samples': 1028,
 'sample_selection_seed': 2023,
 'pruner_name': 'MedianPruner',
 'sampler_name': 'TPESampler',
 'objective_name': 'sparse_small',
 'max_perts': 0,
 'attack_misclassified_samples': False,
 'objective_extra_kwargs': {},
 'pruner_kwargs': {},
 'sampler_kwargs': {}})

The `max_num_samples` parameter specifies the number of samples to be considered for attack. However, samples that are misclassified by the target model are not attacked, so the actual number of samples used for tuning will be slightly lower.

### 9.2 Adversarial Example Quality Scores

Each trial in an attack tuning study runs `epochs_per_batch` attack iterations on each of the selected samples. Then, an adversarial example quality score is calculated for the lowest loss adversarial example each sample that has at least one associated adversarial example. A trial's score is the sum of these example quality scores. The `objective_name` parameter specifies the objective function used to calculate the quality scores.
 
 The following table summarizes how each of the available objective functions calculates the quality score of a single adversarial perturbation matrix $P_{adv}$.

| Objective               | Example Quality Score Formula                       |
| ------------------------| ----------------------------------------------------|
| sparsity                | $1 - f_{nonzero}$                                   |
| max_num_nonzero_perts   | $if\; n_{nonzero} < n_{critical}: 1, otherwise: 0$  |
| sparse_small            | $sparsity\;/\;\|\|P_{adv}\|\|_1$                    |
| sparse_small_max        | $sparsity\;/\ max(\|P_{adv}\|)$                     |

where:
- $n_{nonzero}$ is the number of non-zero elements in $P_{adv}$
- $f_{nonzero}$ is the fraction of non-zero elements
- $\|P_{adv}\|_1$ is the L1 norm
- $|P_{adv}|$ is the element-wise absolute value.

### 9.3 Tune Attack with `sparse_small` Objective

Although, `attack.tuner_driver_settings.objective_name` should already be set to `sparse_small_max`, let's set it again to be sure:


In [51]:
nh.set_config_value("attack.tuner_driver_settings.objective_name", "sparse_small")

We can start a new attack hyperparameter tuning session with the `attack` sub-package's `tune_attack_new` module. If a tuning session is stopped early (via CTRL-C or the notebook Stop button), data from completed trials will be saved.

In [52]:
!python attack/tune_attack_new.py --help

usage: tune_attack_new.py [-h] [-t [CV_TRAINING_ID]] [-r]

Starts a new study for tuning attack hyperparameters.

options:
  -h, --help            show this help message and exit
  -t [CV_TRAINING_ID], --cv_training_id [CV_TRAINING_ID]
                        ID of cross validation tuning session providing
                        trained model for attack.
  -r, --redirect        Redirect terminal output to log file


In [53]:
!python attack/tune_attack_new.py -t {cv_training_id_for_attack_tuning.session_id} -r

Starting new Attack Hyperparameter Tuning session 20250304150536178383.
Using trained model from CV training session ID 20250225120514809962 as the attack target.

Attack tuner driver settings:
{'db_env_var_name': 'ATTACK_TUNING_DB_NAME',
 'epochs_per_batch': 1000,
 'max_num_samples': 1028,
 'max_perts': 0,
 'num_trials': 75,
 'objective_name': 'sparse_small',
 'pruner_kwargs': {},
 'pruner_name': 'MedianPruner',
 'sample_selection_seed': 2023,
 'sampler_kwargs': {},
 'sampler_name': 'TPESampler'}

[I 2025-03-04 15:05:51,484] A new study created in RDB with name: attack_tuning_20250304150536178383

stdout and stderr will be redirected to /home/devspace/project/logs/attack/tuning/20250304150536178383.log
Output can be viewed in real time by running the following command in another terminal:
tail -f /home/devspace/project/logs/attack/tuning/20250304150536178383.log


When attack tuning is complete, store the session.

In [57]:
pipeline_info.store_session(session_type=nh.SessionType.ATTACK_TUNING)

20250305184453643583 stored


### 9.5 Tune Attack with `sparse_small_max` Objective

We change the objective name in our `config.toml` using:

In [55]:
nh.set_config_value("attack.tuner_driver_settings.objective_name", "sparse_small_max")

In [56]:
!python attack/tune_attack_new.py -t {cv_training_id_for_attack_tuning.session_id} -r

Starting new Attack Hyperparameter Tuning session 20250305184453643583.
Using trained model from CV training session ID 20250225120514809962 as the attack target.

Attack tuner driver settings:
{'db_env_var_name': 'ATTACK_TUNING_DB_NAME',
 'epochs_per_batch': 1000,
 'max_num_samples': 1028,
 'max_perts': 0,
 'num_trials': 75,
 'objective_name': 'sparse_small_max',
 'pruner_kwargs': {},
 'pruner_name': 'MedianPruner',
 'sample_selection_seed': 2023,
 'sampler_kwargs': {},
 'sampler_name': 'TPESampler'}

[I 2025-03-05 18:45:09,148] A new study created in RDB with name: attack_tuning_20250305184453643583

stdout and stderr will be redirected to /home/devspace/project/logs/attack/tuning/20250305184453643583.log
Output can be viewed in real time by running the following command in another terminal:
tail -f /home/devspace/project/logs/attack/tuning/20250305184453643583.log


We store this attack tuning session once it is complete.

In [ ]:
pipeline_info.store_session(session_type=nh.SessionType.ATTACK_TUNING)

### 9.5 Resuming an Existing Attack Tuning Session
The `tune_attack_resume` module can be used to run more trials as part of an existing attack tuning study. If we are resuming the most recently created attack tuning study, we can set `custom_attack_tuning_id_to_resume` in the next cell to `None`. Otherwise, set it to the ID of the session we want to resume.

In [50]:
pipeline_info.sessions

{'20250223131513840062': SessionInfo(session_type=<SessionType.DB_QUERIES: 1>, session_id='20250223131513840062', comment=None),
 '20250223131653849429': SessionInfo(session_type=<SessionType.PREPROCESS: 2>, session_id='20250223131653849429', comment=None),
 '20250223132531419387': SessionInfo(session_type=<SessionType.MODEL_TUNING: 3>, session_id='20250223132531419387', comment=None),
 '20250225120514809962': SessionInfo(session_type=<SessionType.CV_TRAINING: 4>, session_id='20250225120514809962', comment=None)}

In [ ]:
# TODO: set this to ID of session to resume (OK to leave as None if resuming most recently created study)
custom_attack_tuning_id_to_resume = None

In [ ]:
attack_tuning_id_to_resume = get_session_id(custom_attack_tuning_id_to_resume, "attack.tuner_driver.output_dir")
attack_tuning_id_to_resume

In [ ]:
!python attack/tune_attack_resume.py --help

In [ ]:
!python attack/tune_attack_resume.py -t {attack_tuning_id_to_resume} -r

## 10. Attacking the Full Dataset with Tuned Attack Hyperparameters

In [ ]:
CONFIG_READER.get_value("attack.driver_settings")

In [ ]:
CONFIG_MODIFIER.set("attack.driver_settings", {'epochs_per_batch': 1000,
 'max_num_samples': 40000,
 'sample_selection_seed': 2023,
 'attack_misclassified_samples': False})

### 10.1 Attack Using Best Hyperparameters from `sparse_small_max` Tuning

In [ ]:
!python attack/attack.py -t {sparse_small_max_attack_tuning_id} -r

## 10.2 Attack Using Best Hyperparameters from `sparse_small` Tuning

In [ ]:
!python attack/attack.py -t {sparse_small_attack_tuning_id} -r

## References

<a id="ref_01"></a>1. [Sun, M., Tang, F., Yi, J., Wang, F. and Zhou, J., 2018, July. Identify susceptible locations in medical records via adversarial attacks on deep predictive models. In *Proceedings of the 24th ACM SIGKDD international conference on knowledge discovery & data mining* (pp. 793-801).](https://dl.acm.org/doi/10.1145/3219819.3219909)

<a id="ref_02"></a>2. [Tang, F., Xiao, C., Wang, F. and Zhou, J., 2018. Predictive modeling in urgent care: a comparative study of machine learning approaches. *Jamia Open*, *1*(1), pp.87-98.](https://academic.oup.com/jamiaopen/article/1/1/87/5032901)

<a><a id="ref_03"></a>3. [Johnson, A., Pollard, T., and Mark, R. (2016) 'MIMIC-III Clinical Database' (version 1.4), *PhysioNet*.](https://doi.org/10.13026/C2XW26) 

<a id="ref_04"></a>4. [Johnson, A. E. W., Pollard, T. J., Shen, L., Lehman, L. H., Feng, M., Ghassemi, M., Moody, B., Szolovits, P., Celi, L. A., & Mark, R. G. (2016). MIMIC-III, a freely accessible critical care database. Scientific Data, 3, 160035.](https://www.nature.com/articles/sdata201635)
